### DISCUSSION TITLE
Word functions

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#91] [Q&A]
If you don't object, I'd start with the word functions. 
I took a closer look at the details and think we need 2 new super instructions:

fndblnk position,string,string-position       find first blank character after string[string-position+1]
fndnblnk position,string,string-position    find first non blank character after string[string-position+1]

with them, we should be able to build all of the functions

**Comment by rvjansen** [comment]
Hi Peter,

Please go ahead.

René.

**Comment by adesutherland** [comment]
Sounds good to me. Can I suggest as well "findchar" instruction that
searches for an arbitrary char (int) in the string.

Also can both instructions search from the string's current position?

Also utf has more than 1 space character! Luckily I have a function to
handle that (I think). I will add details when I have a second ... I am
cooking!

**Comment by Peter-Jacob** [comment]
WORDS is the simplest word function, the others might be more of a nightmare: 
```
/*
     Words function
 */
          .globals=0
main() .locals=8 .expose=test.main
   load  r1,"The quick brown fox jumps over the lazy dog"
   load  r3,0   /* offset in string */
   load  r4,43  /* length of string */
   load  r6,0   /* word count       */
loop:
   fndnblnk r3,r1,r3   /* find next non blank offset */
   ilt r5,r3,0         /* if -1, nothing found, end search */
   brt break,r5
   inc r6              /* else increase word count */
   load r7,"Word found at "
   say r7              /* output offset */
   itos r3
   say r3
   fndblnk r3,r1,r3    /* from offset find next blank offset */
   ilt r5,r3,0         /* if -1, nothing found, end search   */
   brt break,r5
   bct loop,r4,r3      /* then look for next non blank char  */
break:
   load r7,"Words found "
   say r7
   itos r6
   say r6
   ret
```

Result:
```
C:\Users\PeterJ\CLionProjects\CREXX_F030\cmake-build-debug\interpreter\rxvm.exe test
Word found at
0
Word found at
4
Word found at
10
Word found at
16
Word found at
20
Word found at
26
Word found at
31
Word found at
35
Word found at
40
Words found
9
Process finished with exit code 0

```

**Reply by adesutherland** [reply]
Boom!

I have looked at UTF-8 - the good news is that ASCII x20 is always a valid space in UTF8 meaning that this hex value byte is never used in a multibyte character (by clever design). So you code will always work.

However there are about 20 other space characters defined in UTF :-( and a naïve search would be quite inefficient as they are multibyte and I could not find a simple out of the box solution. The answer for us is quite easy - we can use re2c (which we already use for our lexer generator) to create a space search function. However there is no rush, we will get to do this sometime in the future, and meantime we will use your function, is that OK?

BTW I am assuming REXX word() does not accept tab characters as whitespace between words?

Finally, in the utf/utf.h file (which is just some public domain) there are functions you may find useful (like case insensitive string searches).

**Reply by Peter-Jacob** [reply]
puh, 20 space char? what we need is something as an ISSPACE function, if you don't have a better idea. As you suggest, let's stay with '20'x for the moment.

**Reply by adesutherland** [reply]
https://en.wikipedia.org/wiki/Whitespace_character

**Reply by Peter-Jacob** [reply]
in principle, I'd like a delimiter parameter in all word functions, in which you can choose what the separator character is. But I guess this is not really kosher.

**Comment by rvjansen** [comment]
http://www.manmrk.net/tutorials/rexx/oorexx/rexxref42/whitespace.html <http://www.manmrk.net/tutorials/rexx/oorexx/rexxref42/whitespace.html>

Yes, tabs are whitespace for oorexx

**Comment by rvjansen** [comment]
in the standards text, word() is based on subword(), which is based on wordindex() which is based on verify(), which is called with !ALLBLANKS. It might be sensible to have verify() as a machine instruction, as most options of datatype() are shorthand for verify(),
and `AllBlanks = ' '`
Compare.rexx: Str1=translate(Str1, ,!AllBlanks,' ')
Compare.rexx: Str2=translate(Str2, ,!AllBlanks,' ')
DataType.rexx:   if pos(left(String,1),!AllBlanks)>0 then return 0
DataType.rexx:   if pos(right(String,1),!AllBlanks)>0 then return 0
DataType.rexx:     if pos(c,!AllBlanks)>0 then do
DataType.rexx: if pos(left(String,1),!AllBlanks)>0 then return 0
DataType.rexx: if pos(right(String,1),!AllBlanks)>0 then return 0
DataType.rexx:   if pos(c,!AllBlanks)>0 then do
Globals.rexx:    !AllBlanks=' '
Strip.rexx: else Unwanted = !AllBlanks
SubWord.rexx:  if pos(right(Output,1),!AllBlanks) = 0 then leave
WordIndex.rexx:   Start = verify(String, !AllBlanks, 'N', Start)  /* Find non-blank */
WordIndex.rexx:   Start = verify(String, !AllBlanks, 'M', Start + 1) /* Find blank  */

**Comment by rvjansen** [comment]
NetRexx only does blanks for words():

```
 /** Find next non-blank.
      <br>Arg1 is string to search.
      <br>Arg2 is where to seach from.
      <br>Returns 0 if no non-blanks found.
       */
   method nextnonblank(s=char[], start=int) constant returns int
     loop i=start to s.length
       if s[i-1]\=' ' then return i
       end i
     return 0       /* none found */
```

**Reply by rvjansen** [reply]
and because NetRexx works with Java which is unicode internally, like we do, it makes sense to just compare to a space character. I think this will turn out ok even when compiled on EBCDIC based OSes

**Comment by rvjansen** [comment]
As discussed last meeting, I have pushed a branch called feature/f0032 which contains the Rexx source of the Rexx BIFS from the standard. This might be useful for planning and implementation of the Rexx BIFS in Rexx.

**Reply by adesutherland** [reply]
> As discussed last meeting, I have pushed a branch called feature/f0032 which contains the Rexx source of the Rexx BIFS from the standard. This might be useful for planning and implementation of the Rexx BIFS in Rexx.

Is f0032 ready to be merged into develop? (Double checking)

**Comment by Peter-Jacob** [comment]
and here is the WORD function, much easier than expected, not fully tested yet

```
            .globals=0
main()  .locals=11
    /*       0123456789  */
   load  r1,"The quick brown fox jumps over the lazy dog"
   load  r2,2    /* word number */
   load  r3,0    /* offset in string */
   load  r4,43   /* length of string */
   load  r6,0    /* word count       */
   load  R10,""  /* extracted word   */
loop:
   fndnblnk r3,r1,r3   /* find next non blank offset */
   ilt r5,r3,0         /* if <0, nothing found, end search */
   brt break,r5
   inc r6              /* else increase word count  */
                       /* offset of word is in R3   */
   copy r8,r3          /* save offset of word       */
   fndblnk r3,r1,r3    /* from offset find next blank offset */
   ieq r7,r6,r2        /* is this the word we ae looking for?*/
   brt wordf,r7        /* go and fetch it                    */
   ilt r5,r3,0         /* if <0, nothing found, end search   */
   brt break,r5        /* word not found                     */
   bct loop,r4,r3      /* continue to look for next non blank char  */
wordf:
   igt r5,r3,0         /* if <0, make it positiv  */
   brt isplus,r5
   imult r3,r3,-1
isplus:
   isub r5,r3,r8     /* length */
wordc:
   concchar r10,r1,r8
   bct  wordc,r5,r8
   say r10
break:
   ret r10
```

extracting the second word returns:

```
C:\Users\PeterJ\CLionProjects\CREXX_F030\cmake-build-debug\interpreter\rxvm.exe word
quick

Process finished with exit code 0

```

**Comment by rvjansen** [comment]
As it is only documenting the BIF’s using the source as-is, yes, it is ready.
